File: app/src/main/java/com/vladislav/workoutassistant/data/DataRepository.java
package com.vladislav.workoutassistant.data;

import android.app.Application;

import com.vladislav.workoutassistant.R;
import com.vladislav.workoutassistant.data.db.LocalDatabase;
import com.vladislav.workoutassistant.data.db.entity.DiaryEntry;
import com.vladislav.workoutassistant.data.db.entity.Exercise;
import com.vladislav.workoutassistant.data.db.entity.Workout;
import com.vladislav.workoutassistant.data.models.NamedObject;
import com.vladislav.workoutassistant.data.models.WorkoutProgram;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

public class DataRepository {

    private static DataRepository sInstance;

    private Executor mExecutor;
    private LocalDatabase mDatabase;
    private MutableLiveData<DiaryEntry> mTempDiaryEntry;
    private MediatorLiveData<List<DiaryEntry>> mObservableEntries;
    private List<NamedObject> mCategories;
    private List<NamedObject> mIntensityLevels;
    private List<NamedObject> mMuscleGroups;

    private DataRepository(Application application) {
        mDatabase = LocalDatabase.getInstance(application);
        mObservableEntries = new MediatorLiveData<>();

        mObservableEntries.addSource(mDatabase.diaryDao().loadAllEntries(), new Observer<List<DiaryEntry>>() {
            @Override
            public void onChanged(List<DiaryEntry> diaryEntryEntities) {
                if (mDatabase.getDatabaseCreated().getValue() != null) {
                    mObservableEntries.postValue(diaryEntryEntities);
                }
            }
        });

        mCategories = new ArrayList<>();
        String[] names = application.getResources().getStringArray(R.array.categories);
        for (int i = 0; i < names.length; i++) {
            mCategories.add(new NamedObject(i, names[i]));
        }

        mIntensityLevels = new ArrayList<>();
        names = application.getResources().getStringArray(R.array.intensity_levels);
        for (int i = 0; i < names.length; i++) {
            mIntensityLevels.add(new NamedObject(i, names[i]));
        }

        mMuscleGroups = new ArrayList<>();
        names = application.getResources().getStringArray(R.array.muscle_groups);
        for (int i = 0; i < names.length; i++) {
            mMuscleGroups.add(new NamedObject(i, names[i]));
        }

        mTempDiaryEntry = new MutableLiveData<>();
        mTempDiaryEntry.setValue(new DiaryEntry());
        mExecutor = Executors.newSingleThreadExecutor();
    }

    public static DataRepository getInstance(Application application) {
        if (sInstance == null) {
            synchronized (DataRepository.class) {
                if (sInstance == null) {
                    sInstance = new DataRepository(application);
                }
            }
        }
        return sInstance;
    }

    public List<NamedObject> loadCategories() {
        return mCategories;
    }

    public List<NamedObject> loadIntensityLevels() {
        return mIntensityLevels;
    }

    public List<NamedObject> loadMuscleGroups() {
        return mMuscleGroups;
    }

    public LiveData<List<Workout>> loadWorkouts(int categoryId) {
        return mDatabase.workoutDao().loadWorkouts(categoryId);
    }

    public LiveData<Exercise> loadExercise(int exerciseId) {
        return mDatabase.exerciseDao().loadExercise(exerciseId);
    }

    public LiveData<List<Exercise>> loadExercises(List<Integer> exerciseIds) {
        return mDatabase.exerciseDao().loadExercises(exerciseIds);
    }

    public LiveData<List<Exercise>> loadExercises(int muscleGroupId) {
        return mDatabase.exerciseDao().loadExercises(muscleGroupId);
    }

    public LiveData<WorkoutProgram> loadWorkoutProgram(int workoutId) {
        return mDatabase.workoutDao().loadWorkoutProgram(workoutId);
    }

    public LiveData<List<DiaryEntry>> loadDiaryEntries() {
        return mObservableEntries;
    }

    public LiveData<DiaryEntry> loadDiaryEntry(int diaryEntryId) {
        return mDatabase.diaryDao().loadEntry(diaryEntryId);
    }

    public LiveData<DiaryEntry> getNewEntry() {
        DiaryEntry entry = mTempDiaryEntry.getValue();
        entry.setDate(new Date());
        entry.setStartTime(null);
        entry.setFinishTime(null);
        entry.setDuration(null);
        entry.setDefaultTitleChecked(false);
        entry.setTitle(null);
        if (entry.getMuscleGroupsIds().size() > 0) {
            entry.setMuscleGroupsIds(new ArrayList<Integer>());
        }
        return mTempDiaryEntry;
    }

    public void insertNewEntry(final DiaryEntry diaryEntry) {
        mExecutor.execute(new Runnable() {
            @Override
            public void run() {
                mDatabase.diaryDao().insertEntry(diaryEntry);
            }
        });
    }

    public void updateEntry(final DiaryEntry diaryEntry) {
        mExecutor.execute(new Runnable() {
            @Override
            public void run() {
                mDatabase.diaryDao().updateEntry(diaryEntry);
            }
        });
    }

    public void deleteEntriesById(final List<Integer> ids) {
        mExecutor.execute(new Runnable() {
            @Override
            public void run() {
                mDatabase.diaryDao().deleteEntriesById(ids);
            }
        });
    }
}

File: app/src/main/java/com/vladislav/workoutassistant/data/DataRepository.kt
package com.vladislav.workoutassistant.data

import android.app.Application

import com.vladislav.workoutassistant.R
import com.vladislav.workoutassistant.data.db.LocalDatabase
import com.vladislav.workoutassistant.data.db.entity.DiaryEntry
import com.vladislav.workoutassistant.data.db.entity.Exercise
import com.vladislav.workoutassistant.data.db.entity.Workout
import com.vladislav.workoutassistant.data.models.NamedObject
import com.vladislav.workoutassistant.data.models.WorkoutProgram

import java.util.ArrayList
import java.util.Date
import java.util.concurrent.Executor
import java.util.concurrent.Executors

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer

class DataRepository private constructor(application: Application) {

    private val mExecutor: Executor
    private val mDatabase: LocalDatabase = LocalDatabase.getInstance(application)
    private val mObservableEntries: MediatorLiveData<List<DiaryEntry>> = MediatorLiveData()
    private val mCategories: MutableList<NamedObject>
    private val mIntensityLevels: MutableList<NamedObject>
    private val mMuscleGroups: MutableList<NamedObject>

    val newEntry: DiaryEntry
        get() = DiaryEntry()

    init {
        mObservableEntries.addSource(mDatabase.diaryDao().loadAllEntries()) { diaryEntryEntities ->
            if (diaryEntryEntities != null) {
                mObservableEntries.postValue(diaryEntryEntities)
            }
        }

        mCategories = ArrayList()
        var names = application.resources.getStringArray(R.array.categories)
        for (i in names.indices) {
            mCategories.add(NamedObject(i, names[i]))
        }

        mIntensityLevels = ArrayList()
        names = application.resources.getStringArray(R.array.intensity_levels)
        for (i in names.indices) {
            mIntensityLevels.add(NamedObject(i, names[i]))
        }

        mMuscleGroups = ArrayList()
        names = application.resources.getStringArray(R.array.muscle_groups)
        for (i in names.indices) {
            mMuscleGroups.add(NamedObject(i, names[i]))
        }

        mExecutor = Executors.newSingleThreadExecutor()
    }

    fun loadCategories(): List<NamedObject> {
        return mCategories
    }

    fun loadIntensityLevels(): List<NamedObject> {
        return mIntensityLevels
    }

    fun loadMuscleGroups(): List<NamedObject> {
        return mMuscleGroups
    }

    fun loadWorkouts(categoryId: Int): LiveData<List<Workout>> {
        return mDatabase.workoutDao().loadWorkouts(categoryId)
    }

    fun loadExercise(exerciseId: Int): LiveData<Exercise> {
        return mDatabase.exerciseDao().loadExercise(exerciseId)
    }

    fun loadExercises(exerciseIds: List<Int>): LiveData<List<Exercise>> {
        return mDatabase.exerciseDao().loadExercises(exerciseIds)
    }

    fun loadExercises(muscleGroupId: Int): LiveData<List<Exercise>> {
        return mDatabase.exerciseDao().loadExercises(muscleGroupId)
    }

    fun loadWorkoutProgram(workoutId: Int): LiveData<WorkoutProgram> {
        return mDatabase.workoutDao().loadWorkoutProgram(workoutId)
    }

    fun loadDiaryEntries(): LiveData<List<DiaryEntry>> {
        return mObservableEntries
    }

    fun loadDiaryEntry(diaryEntryId: Int): LiveData<DiaryEntry> {
        return mDatabase.diaryDao().loadEntry(diaryEntryId)
    }

    fun insertNewEntry(diaryEntry: DiaryEntry) {
        mExecutor.execute { mDatabase.diaryDao().insertEntry(diaryEntry) }
    }

    fun updateEntry(diaryEntry: DiaryEntry) {
        mExecutor.execute { mDatabase.diaryDao().updateEntry(diaryEntry) }
    }

    fun deleteEntriesById(ids: List<Int>) {
        mExecutor.execute { mDatabase.diaryDao().deleteEntriesById(ids) }
    }

    companion object {

        private var sInstance: DataRepository? = null

        fun getInstance(application: Application): DataRepository {
            if (sInstance == null) {
                synchronized(DataRepository::class.java) {
                    if (sInstance == null) {
                        sInstance = DataRepository(application)
                    }
                }
            }
            return sInstance!!
        }
    }
}